base/utf16.jl

```diff
@@ -1,10 +1,82 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
+#=
+@doc """
+@brief      Error messages for Unicode / UTF support
+""" ->
+=#
+#=
+@enum(UTF_ERR,
+      SHORT = 1,
+      CONT,
+      LONG,
+      NOT_LEAD,
+      NOT_TRAIL,
+      NOT_SURROGATE,
+      MISSING_SURROGATE,
+      INVALID,
+      SURROGATE,
+      NULL_TERMINATE,
+)
+=#
+
+const UTF_ERR_SHORT = 1
+const UTF_ERR_CONT  = 2
+const UTF_ERR_LONG  = 3
+const UTF_ERR_NOT_LEAD = 4
+const UTF_ERR_NOT_TRAIL = 5
+const UTF_ERR_NOT_SURROGATE = 6
+const UTF_ERR_MISSING_SURROGATE = 7
+const UTF_ERR_INVALID = 8
+const UTF_ERR_SURROGATE = 9
+const UTF_ERR_NULL_TERMINATE = 10
+const UTF_ERR_MAX = 10
+
+type UnicodeError <: Exception
+    msg::AbstractString
+end
+
+#=
+@doc """
+@brief      Throws UnicodeError with information about the specific error, location, and character
+
+@param[in]  errCode::UTF_ERR
+@param[in]  errPos:: Integer
+@param[in]  errChar::Integer
+
+@throws never returns, always throws UnicodeError
+""" ->
+=#
+function utf_errfunc(errCode::Integer, errPos::Integer, errChar::Integer)
+    const errMsgs = [
```

Comment:
Part of that comes from the C code, which does not *have* a throw statement.
My C package takes a pointer to an error handler function, which might point to a C++ function, or even a Julia function, which can simply do a throw from within their function if they wish, otherwise, they can simply be a no-op, and return the error code, which is then returned from the C validation function.
Can I pass a variable number of arguments to the exception that I throw? I also remember people telling me not to throw an error, because it might cause a GC frame, so I tried to avoid that in the validation code.
(as a negative number, instead of a 0 or positive number of logical characters).